core/src/main/java/fr/sncf/osrd/envelope/Envelope.java

```diff
@@ -50,4 +84,123 @@ public EnvelopePart get(int i) {
     }
 
     // endregion
+
+    /** Cuts an envelope */
+    public EnvelopePart[] slice(
+            int beginPartIndex, int beginStepIndex, double beginPosition,
+            int endPartIndex, int endStepIndex, double endPosition
+    ) {
+        assert beginPartIndex <= endPartIndex;
+
+        if (beginPartIndex == endPartIndex) {
+            var part = parts[beginPartIndex];
+            var sliced = part.slice(beginStepIndex, beginPosition, endStepIndex, endPosition);
+            if (sliced == null)
+                return new EnvelopePart[] {};
+            return new EnvelopePart[] { sliced };
+        }
+
+        var beginPart = parts[beginPartIndex];
+        var endPart = parts[endPartIndex];
+        var beginPartSliced = beginPart.sliceEnd(beginStepIndex, beginPosition);
+        var endPartSliced = endPart.sliceBeginning(endStepIndex, endPosition);
+
+        // compute the number of unchanged envelope parts between sliced parts
+        var copySize = endPartIndex - beginPartIndex + 1 - /* sliced endpoints */ 2;
+
+        // compute the total sliced envelope size
+        var size = copySize;
+        if (beginPartSliced != null)
+            size++;
+        if (endPartSliced != null)
+            size++;
+
+        var res = new EnvelopePart[size];
+
+        int cur = 0;
+        if (beginPartSliced != null)
+            res[cur++] = beginPartSliced;
+
+        var copyStartIndex = beginPartIndex + 1;
+        for (int i = 0; i < copySize; i++)
+            res[cur++] = parts[copyStartIndex + i];
+
+        if (endPartSliced != null)
+            res[cur] = endPartSliced;
+        return res;
+    }
+
+    public EnvelopePart[] slice(EnvelopePosition begin, EnvelopePosition end) {
+        return slice(
+                begin.getPartIndex(), begin.getStepIndex(), begin.getPosition(),
+                end.getPartIndex(), end.getStepIndex(), end.getPosition()
+        );
+    }
+
+    public EnvelopePart[] smartSlice(
```

Comment:
Transform this into 2 methods sliceBegin & sliceEnd as it's done in EnvelopePart